Feature/spaces

therealtod opened the pull request:
The 'spaces' app introduces a new feature that allows basic CRUD operations on "Place" objects.

therealtod committed:
Add user and user profile management

Through the 'accounts' application, we provide a backend capable of managing user acounts registration and activation, user profile editing and user's password updating.

An user can be created by a POST request to the 'user-list'. The creation of a user automaticaly triggers the creation of his corresponding profile (which right now contains information such as: gender, birthdate, location and nationality,and is easily upgradable in the future with new fields), which can be updated with PUT or PATCH requests to 'user-update-profile' endpoint. Furthermore at the end of the registration process, an activation mail will be sent to the user, providing a link that accepts a GET request, which if correctly delivered will set to "True" the is_active flag of the corresponding user object.
Finally a user can change is own password instantiating  a POST request to the 'user-change-password' endpoint.

therealtod committed:
Places can be updated

therealtod committed:
Add user and user profile management

Through the 'accounts' application, we provide a backend capable of managing user accounts registration and activation, user profile editing and user's password updating.

An user can be created by a POST request to the 'user-list'. The creation of a user automatically triggers the creation of his corresponding profile (which right now contains information such as: gender, birth date, location and nationality,and is easily upgradable in the future with new fields), which can be updated with PUT or PATCH requests to 'user-update-profile' endpoint. Furthermore at the end of the registration process, an activation mail will be sent to the user, providing a link that accepts a GET request, which if correctly delivered will set to "True" the is_active flag of the corresponding user object.
Finally a user can change is own password instantiating  a POST request to the 'user-change-password' endpoint.

therealtod committed:
Add spaces management

The code introduces a new model called "Place" that represents a physical
 site that can host events. It features fields that can store general information (name, description and owner) plus an Address field, which is actually a separate model that encapsulates all the infos about the physical location of the Place.

 Common CRUD operations are supported. A specific @detail_route had been created for the update operation to deal with the nested kind of structure of the "Place" entity.

therealtod committed:
moved accounts.permissions to 'api' app

therealtod committed:
Merge remote-tracking branch 'origin/master' into spaces

# Conflicts:
#	accounts/viewsets.py
#	api/urls.py
#	xea_core/settings.py

ssaavedra commented:
Then it should probably be called the 'places', and not 'spaces' app, and their URLs be on the 'place' prefix, not 'spaces' ;-)

ssaavedra commented on api/permissions.py:
```diff
@@ -0,0 +1,19 @@
+from rest_framework import permissions
```
Es buena idea hacer este refactor, pero si esto fuese una pull request por si mismo, ya estaría probablemente listo para mergear.

ssaavedra commented on spaces/models.py:
```diff
@@ -0,0 +1,23 @@
+
+class Address(models.Model):
+    street_name = models.CharField(max_length=140)
+    civic = models.CharField(max_length=20)
```
What's `civic`?

therealtod replied:
It stands for "civic number". Should we be more specific?

ssaavedra replied:
Does that even exist outside Canada?

On the contrary, I think it's too specific. I would just say

```
address_line1 = models.CharField(max_length=70)
address_line2 = models.CharField(max_length=70)
# address_line3 (maybe)
```

therealtod replied:
I think I didn't explain it very well. That's what I was referring to: https://en.wikipedia.org/wiki/House_numbering

ssaavedra replied:
And, as that very page reads, it seems not every country uses the same numbering system, and many of them are not like ours. And no two countries call their numbers the same uses them in the same way, or are even numbers :-)

ssaavedra commented on spaces/models.py:
```diff
@@ -0,0 +1,23 @@
+    nation = models.CharField(max_length=140)
+
+
+class Place(models.Model):
```
A Place needs GPS coordinates. We need to be able to put it on a map. At least, we need latitude and longitude to draw a point on a map.

ssaavedra commented on spaces/tests.py:
```diff
@@ -0,0 +1,95 @@
+from django.contrib.auth import get_user_model
+from django.core.urlresolvers import reverse
```
`from rest_framework.reverse import reverse`

therealtod commented on spaces/viewsets.py:
```diff
@@ -0,0 +1,31 @@
+    queryset = Place.objects.all()
+    serializer_class = PlaceSerializer
+
+    @detail_route(methods=['put','patch'], permission_classes=[IsPlaceOwnerOrStaff])
```
Would it be "cleaner" to just reimplement the "update" method instead of doing this?

ssaavedra commented on spaces/models.py:
```diff
@@ -0,0 +1,23 @@
+from django.contrib.auth.models import User
+
+
+class Address(models.Model):
```
Yo creo que city debería ser su propia entidad y aquí una ForeignKey, pero eso está abierto a discusión.
[Por favor responded a este comentario discutiéndolo]

ssaavedra replied:
cc/  @gpul-org/xea-devs

ssaavedra commented on spaces/models.py:
```diff
@@ -0,0 +1,23 @@
+    city = models.CharField(max_length=140)
+    postal_code = models.IntegerField(blank=True)
+    region = models.CharField(max_length=140)
+    nation = models.CharField(max_length=140)
```
Nation should just be a two-character field meaning its ISO-3166 code.

ssaavedra commented on spaces/serializers.py:
```diff
@@ -0,0 +1,47 @@
+        model = Place
+        fields = ('url', 'name', 'owner', 'description', 'address')
+
+    def update(self, instance, validated_data):
```
According to http://www.django-rest-framework.org/api-guide/serializers/#modelserializer

The ModelSerializer class is the same as a regular Serializer class, except that:
- [...]
- **It includes simple default implementations of .create() and .update().**

You don't seem to be doing anything that the ModelSerializer would not already do for you.

ssaavedra commented on spaces/serializers.py:
```diff
@@ -0,0 +1,47 @@
+
+    def create(self, validated_data):
+        address_data = validated_data.get('address')
+        address = Address.objects.create(street_name=address_data.get('street_name'),
```
You should never do this. You are replicating code from `address`, and it should not be the PlaceSerializer's responsability.
Just do `address = self.address.create(self, address_data)`

However, as I say in the previous note, this should already be handled by the HyperlinkedModelSerializer and this code might as well not exist.

therealtod replied:
It seems working without the explicit implementation of .update(). So we can immediately remove that code.
However, according to my experience, if we don't explicitly implement the .create() method we run into the following error:

`AssertionError: The`.create()`method does not support writable nested fields by default.
Write an explicit`.create()`method for serializer`spaces.serializers.PlaceSerializer`, or set`read_only=True`on nested serializer fields.`

ssaavedra replied:
Yup. It seems it was me who was mistaken.

ssaavedra replied:
Then, it would be nice if both save's were performed in a single DB transaction (in case one fails so do the others).

For that, we can use [django.db.transaction.atomic()](https://docs.djangoproject.com/en/1.10/topics/db/transactions/#django.db.transaction.atomic) either as a decorator on the update/create function, or as a context inside, as

``` python
with transaction.atomic():
  # stuff here
```

ssaavedra commented on spaces/urls.py:
```diff
@@ -0,0 +1,13 @@
+router.register(r'', PlaceViewSet)
+
+
+urlpatterns = [
```
If this is so simple, you can just do `urlpatterns = router.urls`.

therealtod committed:
Add spaces management

The code introduces a new model called "Place" that represents a physical
 site that can host events. It features fields that can store general information (name, description and owner) plus an Address field, which is actually a separate model

 Common CRUD operations are supported. A specific @detail_route had been created for the update operation to deal with the nested kind of structure of the "Place" entity.

ssaavedra commented on places/serializers.py:
```diff
@@ -20,21 +20,22 @@ def update(self, instance, validated_data):
         instance.name = validated_data.get('name', instance.name)
         instance.owner = validated_data.get('owner', instance.owner)
         instance.description = validated_data.get('nationality', instance.description)
```
description is validated_data.get('nationality') ?

ssaavedra commented on places/models.py:
```diff
@@ -0,0 +1,19 @@
+from django.contrib.auth.models import User
+
+
+class Address(models.Model):
```
Tal vez no he sido claro en absoluto sobre como puede ser interesante manejar las direcciones. Si ha sido así, no era mi intención, y espero aclararlo un poco más con la captura de alguien que, en mi opinión, hace un trabajo relativamente bueno al respecto: Amazon (digo bueno porque, en fin, ellos tienen que realmente mandar paquetes, con lo que saben si estos datos son suficientes o no para llegar al usuario).

De aquí, evidentemente el nombre y apellidos se cambia por el nombre del lugar, que no pertenece a la dirección.

Lo que decía además es que es interesante tener un modelo de ciudades fuera de la dirección, ya que ciudad y país suelen ir juntos, y para poder tener más fácil una consulta del tipo "¿cuántos usuarios tengo registrados que sean de Santiago"? (¿Santiago... de Chile o de España? Bueno, si la ciudad tiene su propio modelo, lo que buscas es aquellos objetos con la clave foránea 27). Pero vamos, que eso no tengo una opinión fuerte al respecto.

Lo que sí opino es:
- Que el país debería de ser el código de dos letras del mismo (esto es un API, queremos datos normalizados, _si se puede_)
- Que el número/letra/puerta/escalera/bloque/punto kilométrico y otras nociones de situación del edificio y la vivienda dentro de un municipio son demasiado dependientes de las costumbres del lugar

Para lo demás, me gustaría poder _hablar_ via comentarios por aquí mismamente al respecto, y poder llegar a algo razonable.

![screenshot from 2016-09-09 15-27-11](https://cloud.githubusercontent.com/assets/581152/18388450/fc9409a6-76a1-11e6-8366-87d16f7a571b.png)

ssaavedra commented:
Ping?

therealtod commented:
Pong:

Esta semana sigo muy líado. Espero poder implementar los cambios que discutimos para la siguiente (o mañana, si acabo las prácticas con antelación)
